### PR TITLE
[Don't Merge] TG-634 String refinement overhaul

### PR DESCRIPTION
**Pull request by LAJW**
A broad collection of ideas to make string solver more comprehensible.

Notable modifications:

 - `axt` - "alternate expression interface" - wrapper around expression allowing us to use C++ operators to build trees of logical expressions.
 - `environmentt` - language environment / factory class for creating `exprt`s from C++ objects in any language.
 - `expr_cast` - Type conversion template for casting `exprt` to any type including strings, integers and types derived from `exprt`. Uses empty optional to signal an error.
 - Turn all private `string_refinement` member functions into free static functions.
 - `string_constraint`s are now regular structs, function definitions relevant to them have been moved to their own module.

This is an initial draft and has some obvious linter issues.

@romainbrenguier Please take a look.

Edit: I am to split this PR into smaller ones:

 1. [Functional style](https://github.com/diffblue/cbmc/pull/1404)
 2. Turn `string_constraintt` into struct
 3. `expr_cast`
 4. `axt`

### CODE REVIEW

**Comment by LAJW** [comment]
@jasigal GL HF

File: src/solvers/refinement/string_refinement.cpp [hunk 1]
```diff
@@ -537,38 +686,78 @@ decision_proceduret::resultt string_refinementt::dec_solve()
...
+        add_lemma(lemma);
       debug() << "check_SAT: got SAT but the model is not correct" << eom;
     }
     else
     {
       debug() << "check_SAT: the model is correct" << eom;
-      concretize_lengths();
+      concretize_lengths(
+        found_length,
+        [](const exprt& expr){ return expr; },
```
Comment (jasigal) [inline src/solvers/refinement/string_refinement.cpp]: Should probably be `this->get` or similar, not identity.

File: src/solvers/refinement/string_refinement.cpp [hunk 2]
```diff
@@ -537,38 +686,78 @@ decision_proceduret::resultt string_refinementt::dec_solve()
...
+          add_lemma(lemma);
         debug() << "check_SAT: got SAT but the model is not correct" << eom;
       }
       else
       {
         debug() << "check_SAT: the model is correct" << eom;
-        concretize_lengths();
+        concretize_lengths(
+          found_length,
+          [](const exprt& expr){ return expr; },
```
Comment (jasigal) [inline src/solvers/refinement/string_refinement.cpp]: Should probably be `this->get` or similar, not identity.

**Comment by tautschnig** [comment]
What's the state of affairs of this work? Quite a lot has happened to the string solver in the meantime...

**Comment by romainbrenguier** [comment]
@tautschnig several parts of this work have been isolated in individual PRs. This will be rebased once the other parts are merged.

**Comment by romainbrenguier** [comment]
@LAJW 
I rebased this branch on develop:
https://github.com/diffblue/cbmc/compare/develop...romainbrenguier:string-refinement-overhaul

Should I replace this PR by a new one, or force push on this one?

**Comment by LAJW** [comment]
This was reduced to a tiny refactor and would have to be reimplemented anyway, because rebasing no longer makes any sense. Closing.

**Pull request closed without merging**